content/blog/2019-09-25-exceptions-in-bril.md

```diff
@@ -0,0 +1,155 @@
++++
+title = "Exceptions in Bril"
+extra.author = "Rolph Recto"
+extra.bio = """
+[Rolph Recto](rolph-recto.github.io) is a third-year graduate student studying
+the intersection of programming languages, security, and distributed systems.
+"""
++++
+
+
+## Overview
+
+We add exceptions as a control structure in Bril to support non-local
+control flow. 
+This allows frontends to compile exceptions directly.
+The ultimate goal is to extend Bril further with more powerful control
+structures such as first-class continuations or algebraic effects.
+
+
+## Function Calls
+
+To make the implementation of exceptions nontrivial and to make exceptions
+actually be non-local control structures, we first extend Bril so that it
+supports function calls.
+We do this by adding a `call` instruction that takes as arguments a
+function name and a list of function arguments.
+We also extended function declarations to allow for formal parameters.
+The Bril text format now allows for declarations of the form
+
+```
+funcName arg1:type1 ... argN:typeN { instrs }
+```
+
+Our implemention explicitly represents the program stack as an object. 
+
+The interpreter has four configuration variables that consist of its current
+activation record:
+
+* Bril function object --- the current function whose list of instructions
+  is being executed
+
+* Variable environment --- map from local variables to values
+
+* Handler environment --- map from exception handlers to (see below)
+
+* PC index --- the index into the instruction list of the current instruction
+  being executed
+
+The interpreter proceeds by executing instructions, which either change
+environment mappings and/or sets the PC index of the next instruction
+to execute.
+When `call` is executed, the interpreter does the following:
+
+* The current activation record is pushed into the program stack.
+
+* An empty variable environment is constructed, and the formal parameters
+  of the function are mapped to the arguments of the `call` instruction.
+  An empty handler environment is constructed.
+
+* The PC index is changed to 0, and the variable and handler environments 
+  are updated.
```

Comment:
How are the variable and handler environments updated? (I thought we just initialized them in the previous step?)

Reply:
I meant that the environments were emptied since the function call creates a new activation record.